### ISSUE TITLE
Proposal: Union types should emit metadata arrays

### ISSUE THREAD

**Issue by Neme** [Microsoft/TypeScript#9916]
# Problem

Currently, metadata type information is not emitted on union types.  When you try to get metadata type information of union types you always get an **Object**.
Returning an array of types would be more useful.

# Proposal

Union types should emit metadata array.

# Examples

## Union types

```ts
function Decorator(constructor: Function) {}

@Decorator
class Foo {
    constructor(param: number | string) {}
}
```

_~>Compiled_

**Currently**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [Object])
    ], Foo);
```

**Proposal**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [[Number, String]])
    ], Foo);
```

## String Literal Types

```ts

function Decorator(constructor: Function) { }

@Decorator
class Foo {
    constructor(param: 'North' | 'East' | 'South' | 'West') { }
}
```

_~>Compiled_

**Currently**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [Object])
    ], Foo);
```

**Proposal**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [[ 'North', 'East', 'South', 'West']])
    ], Foo);
```

## Null
```ts
function Decorator(constructor: Function) { }

@Decorator
class Foo {
    constructor(param: string | null) { }
}
```

_~>Compiled_

**Currently**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [Object])
    ], Foo);
```

**Proposal**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [[String,  null]])
    ], Foo);
```
## Undefined
```ts
function Decorator(constructor: Function) { }

@Decorator
class Foo {
    constructor(param: string | undefined) { }
}
```

_~>Compiled_

**Currently**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [Object])
    ], Foo);
```

**Proposal**
```js
Foo = __decorate([
        Decorator, 
        __metadata('design:paramtypes', [String])
    ], Foo);
```

# Use case

For example, when you want the type information to generate your database tables.
```ts
class Age {

    @Attribute
    public name :string;   // VARCHAR
     

    @Attribute
    public gender : ['male', 'female'];   // ENUM('male', 'female')
}
```

**Comment by mhegazy** [comment]
This is another variation of the request in https://github.com/Microsoft/TypeScript/issues/3015. The decorator metadata feature is not meant to provide a full runtime type system, or a reflection-like feature.  Please see my response in https://github.com/Microsoft/TypeScript/issues/3015#issuecomment-128149650

**Comment by Neme** [comment]
I would love to see a feature like #3015 in typescript.
This would give us the opportunity for so many things.

If the .js file gets to big why don't support a specific metafile like [https://github.com/artifacthealth/tsreflect-compiler](here) natively.
A compiler option (like --metaFile) would generete such a .json file.

**Comment by mhegazy** [comment]
you can not separate type info in one object, as not all types are visible from the top level scope, e.g. an interface inside a function declaration.

so types will need to be maintained some how at run time. the size of the information is not small. more over, the overhead of creating these objects, arrays, and strings is not trivial on the startup cost of an app.

also worth noting that types themselves are not sufficient, you will need a runtime library to query these types, and to check their assingbality the same way the compiler does at design time, and get their comments etc..

in short there is not really a middle ground here, once you start down the runtime-type system, you need all of it, and you end up with a reflection API like in C# for instance.

for all of that, this is out of the scope of the typescript project at the time being.
